cluster-autoscaler/cloudprovider/gce/mig_info_provider.go

```diff
@@ -165,6 +184,98 @@ func (c *cachingMigInfoProvider) RegenerateMigInstancesCache() error {
 	return nil
 }
 
+func (c *cachingMigInfoProvider) listInstancesInAllZonesWithMigs() ([]GceInstance, error) {
+	var zones []string
+	for zone := range c.listAllZonesWithMigs() {
+		zones = append(zones, zone)
+	}
+	var allInstances []GceInstance
+	errors := make([]error, len(zones))
+	zoneInstances := make([][]GceInstance, len(zones))
+	workqueue.ParallelizeUntil(context.Background(), c.concurrentGceRefreshes, len(zones), func(piece int) {
+		zoneInstances[piece], errors[piece] = c.gceClient.FetchAllInstances(c.projectId, zones[piece], "")
+	}, workqueue.WithChunkSize(c.concurrentGceRefreshes))
+
+	for _, instances := range zoneInstances {
+		allInstances = append(allInstances, instances...)
+	}
+	for _, err := range errors {
+		if err != nil {
+			return allInstances, err
+		}
+	}
+	return allInstances, nil
+}
+
+func groupInstancesToMigs(instances []GceInstance) map[GceRef][]GceInstance {
+	migToInstances := map[GceRef][]GceInstance{}
+	for _, instance := range instances {
+		migToInstances[instance.Igm] = append(migToInstances[instance.Igm], instance)
+	}
+	return migToInstances
+}
+
+func (c *cachingMigInfoProvider) isMigInstancesConsistent(mig Mig, migToInstances map[GceRef][]GceInstance) bool {
+	migRef := mig.GceRef()
+	instances, found := migToInstances[migRef]
+	if !found {
+		return false
+	}
+
+	state, found := c.cache.GetMigInstancesState(migRef)
+	if !found {
+		return false
+	}
+
+	instanceCount := state[cloudprovider.InstanceRunning] + state[cloudprovider.InstanceCreating] + state[cloudprovider.InstanceDeleting]
+	return instanceCount == int64(len(instances))
+}
+
+func (c *cachingMigInfoProvider) isMigCreatingOrDeletingInstances(mig Mig) bool {
+	migRef := mig.GceRef()
+	state, found := c.cache.GetMigInstancesState(migRef)
+	if !found {
+		return false
+	}
+	return state[cloudprovider.InstanceCreating] > 0 || state[cloudprovider.InstanceDeleting] > 0
+}
+
+// updateMigInstancesCache updates the mig instances for each mig
+func (c *cachingMigInfoProvider) updateMigInstancesCache(migToInstances map[GceRef][]GceInstance) error {
+	var errors []error
+	for _, mig := range c.migLister.GetMigs() {
+		migRef := mig.GceRef()
+		// mig instances are re-fetched if there is an inconsistency between number
```

Comment:
Maybe: 
If there is an inconsistency between <...> and <...> for the given mig, which can be due to
- abandoned instance
- missing/malformed "created-by" reference

we use an igm.ListInstances call as the authoritative source of instance information

Reply:
Done